R/separate_longer.R

```diff
@@ -0,0 +1,255 @@
+#' df5 <- pl$DataFrame(
+#'   id = 1:4,
+#'   x = c("ab", "", "ef", NA)
+#' )
+#' separate_longer_position_polars(df5, x, width = 2)
+#' separate_longer_position_polars(df5, x, width = 2, keep_empty = TRUE)
+#'
+#' # Multiple columns with broadcasting
+#' df6 <- pl$DataFrame(
+#'   id = 1:3,
+#'   x = c("a", "bc", "def"),
+#'   y = c("12", "345", "67")
+#' )
+#' # Shorter strings are recycled to match the longest in each row
+#' separate_longer_position_polars(df6, c(x, y), width = 2)
+separate_longer_delim_polars <- function(
+  data,
+  cols,
+  delim,
+  ...
+) {
+  check_polars_data(data)
+  rlang::check_dots_empty()
+  rlang::check_required(cols)
+  check_string(delim)
+
+  # Capture the column expression and filter to only string columns
+  col_names <- tidyselect_named_arg(data, rlang::enquo(cols)) |>
+    intersect(names(data$select(pl$col(pl$String))))
+
+  # If no string columns, return data unchanged (like tidyr)
+  if (length(col_names) == 0) {
+    return(add_tidypolars_class(data))
+  }
+
+  # Split each column by delimiter and convert to list
+  out <- data$with_columns(
+    do.call(pl$col, as.list(col_names))$str$split(delim)
+  )
+
+  # Handle multi-column broadcasting and validation
+  if (length(col_names) > 1) {
+    # Convert lists containing only empty strings [""] to empty lists []
+    # This matches tidyr behavior where "" split by "," gives character(0)
+    # Only do this for multi-column case to enable proper broadcasting
+    empty_list_exprs <- lapply(col_names, function(nm) {
+      col <- pl$col(nm)
+      # Check if list has length 1 and that element is ""
+      is_empty_string <- col$list$len()$eq(1)$and(col$list$first()$eq(""))
+      # If so, replace with empty list, otherwise keep original
+      pl$when(is_empty_string)$then(
+        pl$lit(list(character(0)))$cast(pl$List(pl$String))
+      )$otherwise(col)$alias(nm)
+    })
+    out <- out$with_columns(!!!empty_list_exprs)
+
+    out <- handle_multi_column_explode(out, col_names)
+  }
+
+  # Explode all columns together
+  add_tidypolars_class(out$explode(col_names))
+}
+
+
+#' @rdname separate_longer
+#' @export
+separate_longer_position_polars <- function(
+  data,
+  cols,
+  width,
+  ...,
+  keep_empty = FALSE
+) {
+  check_polars_data(data)
+  rlang::check_dots_empty()
+  rlang::check_required(cols)
+  check_number_whole(width, min = 1)
+  check_bool(keep_empty)
+
+  # Capture the column expression and filter to only string columns
+  col_names <- tidyselect_named_arg(data, rlang::enquo(cols)) |>
+    intersect(names(data$select(pl$col(pl$String))))
+
+  # If no string columns, return data unchanged (like tidyr)
+  if (length(col_names) == 0) {
+    return(add_tidypolars_class(data))
+  }
+
+  # Use regex to extract groups of `width` characters
+  # Pattern: .{1,width} matches 1 to width characters (greedy)
+  pattern <- paste0(".{1,", width, "}")
+  out <- data$with_columns(
+    do.call(pl$col, as.list(col_names))$str$extract_all(pattern)
```

Comment:
I think `$extract(pattern, 0)` is identical and possibly less expensive.

It's a bit sad the split functions in polars do not allow to split by width, I guess an alternative could be to chain `str$slice()` but that doesn't necessarily sound more efficient.